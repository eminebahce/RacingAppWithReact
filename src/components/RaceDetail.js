import React, {Component} from 'react';
import {Badge, Button, Popover, PopoverBody, PopoverHeader, Table} from 'reactstrap';
import Moment from 'react-moment';

class RaceDetail extends Component {
    constructor(props) {
        super(props);

        this.state = ({
            buttonIds: []
        });
        this.toggle = this.toggle.bind(this);
    }

    toggle(toggledButtonId) {
        this.setState({
            [toggledButtonId]: !this.state[toggledButtonId]
        });
    }

    componentWillReceiveProps(nextProps) {
        if (Object.keys(nextProps.race).length !== 0) {
            let newButtonIds = [];
            nextProps.race.starts.map((start, index) => {
                newButtonIds[index] = `btn${nextProps.race.id}${start.number}`;
                this.setState({
                    [newButtonIds[index]]: false
                });
            });

            this.setState({
                buttonIds: newButtonIds
            });
        }
    }

    render() {
        let starts = '';
        let raceData = '';
        if (Object.keys(this.props.race).length !== 0) {
            starts = this.props.race.starts.map((start, index) => {
                return (
                    <tr key={this.state.buttonIds[index]}>
                        <td>{start.number}</td>
                        <td>{start.horse.name}</td>
                        <td>{start.driver.firstName} {start.driver.lastName}</td>
                        <td>
                            <div>
                                <Button color="info" id={this.state.buttonIds[index]} onClick={() => this.toggle(this.state.buttonIds[index])}>
                                    +
                                </Button>
                                <Popover placement="bottom" isOpen={this.state[this.state.buttonIds[index]]} target={this.state.buttonIds[index]}
                                         toggle={() => this.toggle(this.state.buttonIds[index])}>
                                    <PopoverHeader>Information</PopoverHeader>
                                    <PopoverBody>
                                        <div>
                                            Trainer: {start.horse.trainer.firstName} {start.horse.trainer.lastName}
                                        </div>
                                        <div>
                                            Horse Father: {start.horse.pedigree.father.name}
                                        </div>
                                    </PopoverBody>
                                </Popover>
                            </div>
                        </td>
                    </tr>
                );
            });

            raceData = (
                <div>
                    <Table size="sm" striped responsive>
                        <thead>
                        <tr>
                            <th colSpan="4" className="tableRowAlign">
                                <Badge color="success"
                                    pill>{this.props.race.number}</Badge> : {this.props.race.name} : <Moment>{this.props.race.startTime}</Moment>
                            </th>
                        </tr>
                        <tr>
                            <th>Start #</th>
                            <th>Horse Name</th>
                            <th>Driver</th>
                            <th>Info</th>
                        </tr>
                        </thead>
                        <tbody>
                        {starts}
                        </tbody>
                    </Table>
                </div>
            );
        }
        return (
            <div>{raceData}</div>
        );
    }
}

export default RaceDetail;