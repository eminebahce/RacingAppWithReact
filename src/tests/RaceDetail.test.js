import React from 'react';
import Renderer from 'react-test-renderer';
import RaceDetail from '../components/RaceDetail';

it('RaceDetail component renders raceDetail correctly', () => {
    let race = {
        id: '1',
        number: 5,
        name: 'Byggnads Prostataloppet - Bronsdivisionen, Final',
        startTime: '2018-05-26T14:30:00',
        starts: [
            {
                number: '1',
                horse: {
                    name: 'Disco Volante',
                    trainer: {
                        firstName: 'Stefan',
                        lastName: 'Melander'
                    },
                    pedigree: {
                        father: {
                            name: 'Scarlet Knight'
                        }
                    }
                },
                driver: {
                    firstName: 'Ulf',
                    lastName: 'Ohlsson'
                }
            }
        ]
    };
    const component = Renderer.create(
        <RaceDetail race={race}/>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});
