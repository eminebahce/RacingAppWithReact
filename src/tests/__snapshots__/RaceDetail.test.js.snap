// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RaceDetail component renders raceDetail correctly 1`] = `
<div>
  <div>
    <div
      className="table-responsive"
    >
      <table
        className="table table-sm table-striped"
      >
        <thead>
          <tr>
            <th
              className="tableRowAlign"
              colSpan="4"
            >
              <span
                className="badge badge-success badge-pill"
              >
                5
              </span>
               : 
              Byggnads Prostataloppet - Bronsdivisionen, Final
               : 
              <time
                dateTime={"2018-05-26T11:30:00.000Z"}
              >
                Sat May 26 2018 14:30:00 GMT+0300
              </time>
            </th>
          </tr>
          <tr>
            <th>
              Start #
            </th>
            <th>
              Horse Name
            </th>
            <th>
              Driver
            </th>
            <th>
              Info
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              1
            </td>
            <td>
              Disco Volante
            </td>
            <td>
              Ulf
               
              Ohlsson
            </td>
            <td>
              <div>
                <button
                  className="btn btn-info"
                  id={undefined}
                  onClick={[Function]}
                  type="button"
                >
                  +
                </button>
              </div>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
